### PR TITLE
Feature: Vim-like Navigation on Cards

### PR DESCRIPTION
**Pull request by sshanzel**
As discussed, selected keys at issue #84 shall be included in the Vim-Like Navigation.

I've applied the hover effect to allow the user to identify which post is focused using the new feature. I applied it by adding a `hover` class to the element which is an existing animation. I only had to move a bit the placements of the styles in the css. 

When you click any of the navigation keys, it would hover the very first post (be it ad or not.) since there is no initial state yet to identify where should it be navigated. I've thought of the top left most post since it seemed sensible.

This also focuses the anchor element in which when you press `enter` the post shall be opened.

**NOTE**: Please checkout the `insane` mode, since there is no animation for hovering yet on the ad, i've added a border effect to temporarily identify it is navigated. And so it will be easier to apply what effect will be implemented after discussing it here, since the functions are already in placed.

### CODE REVIEW

File: packages/extension/src/components/DaFeed.vue [hunk 1]
```diff
@@ -106,6 +122,176 @@ export default {
...
+
+    getFirstChildElement(parent) {
+      const child = Array.from(parent.childNodes).find(child =>
+          (child.previousSibling === null && child.nodeType == Node.ELEMENT_NODE) ||
+          (child.previousSibling && child.previousSibling.nodeType !== Node.ELEMENT_NODE));
+
+      const index = this.getPostElementIndex(child);
+
+      return index >= 0 ? child : child.nextSibling;
+    },
```
Comment (sshanzel) [inline packages/extension/src/components/DaFeed.vue]: Didn't realize there's a `firstElementChild` api. Lol

File: packages/extension/src/components/DaFeed.vue [hunk 2]
```diff
@@ -106,6 +122,176 @@ export default {
...
+    getPostElementIndex(postElement) {
+      return this.$refs.posts.findIndex(post => post.$el == postElement);
+    },
+
+    getElementIndexFromSiblings(targetElement) {
+      let index = 0, element = targetElement;
+      while ((element = element.previousSibling) != null) index++;
+
+      return index;
+    },
```
Comment (sshanzel) [inline packages/extension/src/components/DaFeed.vue]: I shall take comments into consideration on this part as well.

File: packages/extension/src/components/DaFeed.vue (hunk 1 above)
Comment (idoshamun) [reply packages/extension/src/components/DaFeed.vue]: Could you please make sure to use it before I start the review? :pray:

**Comment by idoshamun** [comment]
@sshanzel I appreciate your time and the will to contribute once more 🙌
Regarding your note about the insane mode, there is actually an existing hover mode but only for the `DaInsanePost` not for the ad. Once you hover on an insane post many options are revealed such as the bookmark icon. Try to take a look and to fix your implementation accordingly, please.

**Comment by sshanzel** [comment]
Hi Ido, yes i've implemented it already for the insane mode in this PR, but it is the ad the only thing missing.

**Comment by idoshamun** [comment]
Perfect, thank you! Will review as soon as possible

**Comment by sshanzel** [comment]
Later today, i will also apply the concerns i saw based from my personal review 😄

**Comment by sshanzel** [comment]
It's still lunch and i realized and had the minutes, so i've commited the reviews i commented 😄

**Comment by idoshamun** [comment]
I noticed that my CI didn't enforce the code styling before, so I've just deployed a fix.
Now apparently your PR doesn't comply with our lint, please fix it before we proceed.

**Comment by sshanzel** [comment]
Hi @idoshamun, i have now fixed the linting issues for my created PR. I am now starting to appreciate linting and tests more because this. I'll make sure to create tests on my futures contributions! 😃

**Comment by sshanzel** [comment]
Thank you very much for the wonderful comments, i shall make the changes right away. This is becoming to feel like i'm the one receiving so much by improving my code game with you experience being shared. Can't believe this kind of community exists. Thank you so much all the way!

File: packages/extension/src/components/DaFeed.vue [hunk 3]
```diff
@@ -106,6 +125,162 @@ export default {
...
+      const [insane] = post.ad ? [post.$el] : post.$el.getElementsByClassName('insane--post');
+
+      insane.getElementsByClassName('insane__link')[0].focus();
+      insane.classList.add('hover');
+    },
+
+    animateCard(post) {
+      post.$el.getElementsByClassName('card__link')[0].focus();
+      post.$el.classList.add('hover');
+    },
```
[the remark this answers is not in the document]
Comment (sshanzel) [reply packages/extension/src/components/DaFeed.vue]: Oh yeah i didn't realize this early on. I was actually using this at first then when i applied the bookmark feat, but bookmark rerenders the card which resets the classes and led me to pass the hover property. Guess i shall remove this function to keep the fix for bookmarking intact.

**Comment by sshanzel** [comment]
I am now working on creating the service for keystrokes. Will push ASAP 😁

**Comment by idoshamun** [comment]
Perfect! One thing that I forgot is tests. I think we can add a few tests to make sure everything is working

**Comment by sshanzel** [comment]
Will sure to provide tests cases the best way i can! Though this is gonna be the first time (and i'm excited about it, lol) i'll be creating few so please let me know straight in the future if any issue is seen.

**Comment by idoshamun** [comment]
Awesome 😊 let me know if you need anything

**Comment by sshanzel** [comment]
Hi @idoshamun! I've finished making the changes, i've also prepared few tests as well. Please let me know if there's any concern with the current updates. Hoping for the best!

File: packages/components/src/components/DaCard.vue [hunk 4]
```diff
@@ -226,13 +226,14 @@ export default {
       box-shadow: 0 16px 24px 0 rgba(0, 0, 0, 0.64);
       transition: opacity 0.4s ease-in-out;
     }
+  }
 
-    & .hover, :hover {
-      transform: translateY(-4px);
```
Comment (sshanzel) [inline packages/components/src/components/DaCard.vue]: initial fix was able to restrict animation based on configuration but this style is not capable of performing the animation when allowed.  had to explicitly mention the `.card` selector.

File: packages/components/src/common/domHelper.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+export function hoverPost(selectedPost, insaneMode) {
+  const linkType = insaneMode ? 'insane__link' : 'card__link';
+  
+  selectedPost.$el.getElementsByClassName(linkType)[0].focus();
+}
+
+export default {
+  getElementIndexFromSiblings
+}
```
Comment (idoshamun) [inline packages/components/src/common/domHelper.js]: Please add a new line here

File: packages/components/src/common/domHelper.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  while(!result && child) {
+    child = child.nextElementSibling;
+    result = getPostByElement(posts, child);
+  }
+
+  return child;
+}
+
+export function hoverPost(selectedPost, insaneMode) {
+  const linkType = insaneMode ? 'insane__link' : 'card__link';
```
Comment (idoshamun) [inline packages/components/src/common/domHelper.js]: You can add the same class for both elements to overcome the condition here.
It makes more sense to not care about the link type, in my opinion.

File: packages/components/src/common/adMixin.js [hunk 7]
```diff
@@ -4,6 +4,14 @@ export default {
       type: Object,
       required: true,
     },
+    showMenu: {
+      type: Boolean,
+      default: true,
+    },
```
Comment (idoshamun) [inline packages/components/src/common/adMixin.js]: Why do you need `showMenu` here?

File: packages/components/src/components/DaCard.vue [hunk 8]
```diff
@@ -207,14 +208,7 @@ export default {
...
 
-.card:hover,
-.card.hover {
-  & .card__background__image {
-    transform: translate(0, 4px) scale(1.05);
-  }
-}
-
-.animate-cards {
+.animate-cards .feed {
```
Comment (idoshamun) [inline packages/components/src/components/DaCard.vue]: Why did you add the `.feed` selector? I think it's too specific.
We can use cards potentially everywhere.

File: packages/components/src/components/DaInsaneAd.vue [hunk 9]
```diff
@@ -22,6 +23,14 @@ export default {
   components: {
     DaLineClamp,
   },
+  computed: {
+    cls() {
+      return {
+        hover: this.selected,
+        'hide-menu': !this.showMenu,
```
Comment (idoshamun) [inline packages/components/src/components/DaInsaneAd.vue]: There is no menu for ads

File: packages/extension/src/common/keyNavigationService.js [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+  const posts = feedComp.$refs.posts;
+  const insaneMode = store.state.ui.insaneMode;
+
+  if (posts.length === 0) return null;
+
+  const item = getCurrentPost(posts, current);
+
+  if (Object.values(validKeys).indexOf(keyCode) === -1) return null;
+
+  if (keyCode === validKeys['/']) return feedComp.$parent.enableSearch();
```
Comment (idoshamun) [inline packages/extension/src/common/keyNavigationService.js]: Please don't rely on that the `enableSearch` function is in the feed's parent.
If needed provide another argument to this service.
Relying on the DOM structure can be very sensitive.

File: packages/extension/src/routes/Home.vue [hunk 11]
```diff
@@ -484,6 +491,8 @@ export default {
       await this.initHome();
       await this.validateAuth();
     });
+
+    window.addEventListener('keypress', this.onKeyPress)
```
Comment (idoshamun) [inline packages/extension/src/routes/Home.vue]: Please be sure to remove the event listener when needed as well

File: packages/extension/src/components/DaFeed.vue [hunk 12]
```diff
@@ -54,6 +59,7 @@ export default {
   },
   data() {
     return {
+      selectedPost: null,
```
Comment (idoshamun) [inline packages/extension/src/components/DaFeed.vue]: This name is very confusing with the property below

**Comment by sshanzel** [comment]
Thank you very much again Ido for the kind reviews and patience! Will make sure to rectify the concerns mentioned above.

File: packages/components/src/common/adMixin.js (hunk 7 above)
Comment (sshanzel) [reply packages/components/src/common/adMixin.js]: Oh, i didn't realize this got included. my bad! 😁

File: packages/components/src/components/DaCard.vue (hunk 8 above)
Comment (sshanzel) [reply packages/components/src/components/DaCard.vue]: Oh sorry, i was debugging something and wasn't able to remove after. Thanks for noticing!

File: packages/extension/src/common/keyNavigationService.js (hunk 10 above)
Comment (sshanzel) [reply packages/extension/src/common/keyNavigationService.js]: this is true, i will also be removing the dependency from the store and add an optional argument which can consist of nullable values.

File: packages/components/src/common/domHelper.js (hunk 6 above)
Comment (sshanzel) [reply packages/components/src/common/domHelper.js]: yeah this is true, so silly of me. Lol

File: packages/components/src/components/DaInsanePost.vue [hunk 13]
```diff
@@ -1,7 +1,8 @@
 <template>
   <div class="insane__wrapper">
     <div class="insane insane--post" :class="cls">
-      <a :href="post.url" target="_blank" class="insane__link" @click="$emit('click', post)">
+      <a :href="post.url" target="_blank" class="insane__link post__link"
+          @click="$emit('click', post)">
```
Comment (sshanzel) [inline packages/components/src/components/DaInsanePost.vue]: This is weird, the linting checker was not able to see the latest changes.

**Comment by sshanzel** [comment]
Hi @idoshamun, i've applied the changes now, i was also able to understand the css styling now, i've applied opacity on the menu button for `insanePost` since the effect wasn't taking effect with `display: block`. I'm truly sorry for the troubles! Hoping for the best!

File: packages/components/src/components/DaInsaneAd.vue [hunk 14]
```diff
@@ -1,14 +1,15 @@
 <template>
-  <div class="insane insane--ad">
-    <a :href="ad.link" target="_blank" class="insane__link" @click="$emit('click', ad)">
+  <div class="insane insane--ad hide-menu" :class="cls">
```
Comment (sshanzel) [inline packages/components/src/components/DaInsaneAd.vue]: i had to add a `hide-menu` class, since it pushes to left and leaving extra white space for the menu since menu configuration was not mentioned.

**Comment by idoshamun** [comment]
@sshanzel Let me know once you are done with the PR and I'll merge :)

**Comment by sshanzel** [comment]
Hi @idoshamun, i just finalized with my own review and rectified. I am now 100% confident with this PR. I really appreciate you taking the time to review my contributions. Thank you very much all the way! YOU ARE THE BEST!
